allennlp/layers/embeddings.py

```diff
@@ -0,0 +1,173 @@
+
+    Returns
+    -------
+    An Embedding module.
+
+    """
+
+    def __init__(self,
+                 num_embeddings: int,
+                 embedding_dim: int,
+                 weight: torch.FloatTensor=None,
+                 padding_index: int=None,
+                 trainable: bool=True,
+                 max_norm: float=None,
+                 norm_type: float=2.,
+                 scale_grad_by_freq: bool=False,
+                 sparse: bool=False):
+        super(Embedding, self).__init__()
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = embedding_dim
+        self.padding_index = padding_index
+        self.max_norm = max_norm
+        self.norm_type = norm_type
+        self.scale_grad_by_freq = scale_grad_by_freq
+        self.sparse = sparse
+
+        if embedding_dim == 1:
+            raise ConfigurationError("There is no need to embed tokens if you "
+                                     "are using a 1 dimensional embedding.")
+
+        if weight is None:
+            weight = torch.FloatTensor(num_embeddings, embedding_dim)
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            self.reset_parameters()
+
+        else:
+            if weight.size() != (num_embeddings, embedding_dim):
+                raise ConfigurationError("A weight matrix was passed with contradictory embedding shapes.")
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            if self.padding_index is not None:
+                self.weight.data[self.padding_index].fill_(0)
+
+    def reset_parameters(self):
+        self.weight.data.normal_(0, 1)
+        if self.padding_index is not None:
+            self.weight.data[self.padding_index].fill_(0)
+
+    def forward(self, input):
+        padding_idx = self.padding_idx
+        if padding_idx is None:
+            padding_idx = -1
+        return self._backend.Embedding(padding_idx,
+                                       self.max_norm,
+                                       self.norm_type,
+                                       self.scale_grad_by_freq,
+                                       self.sparse)(input, self.weight)
+
+
+def get_pretrained_embedding_layer(embeddings_filename: str,
+                                   vocab: Vocabulary,
+                                   namespace: str="tokens",
+                                   trainable: bool=True,
+                                   log_misses: bool=False):
+    """
+    Reads a pre-trained embedding file and generates a Keras Embedding layer that has weights
+    initialized to the pre-trained embeddings.  The Embedding layer can either be trainable or
+    not.
+
+    We use the DataIndexer to map from the word strings in the embeddings file to the indices
+    that we need, and to know which words from the embeddings file we can safely ignore.  If we
```

Comment:
We probably don't want to ignore words in the pretrained file that aren't in the vocab, but we can handle that later.